### PR TITLE
feat: Use service manager for transport to be standalone #193

### PR DESCRIPTION
**Pull request by Eywek**
- [x] Exception
- [x] Logs
- [x] Profiling
- [x] Custom metrics
- [x] Custom actions

### CODE REVIEW

File: src/pmx.ts [hunk 1]
```diff
@@ -121,15 +123,33 @@ export default class PMX {
...
+      Configuration.init(config)
+      this.initialConfig = config
+    }
+
+    // Transport
+    if (config.standalone && config.publicKey && config.secretKey) {
+      ServiceManager.get('transport').initStandalone({
+        publicKey: config.publicKey,
+        secretKey: config.secretKey,
+        appName: config.appName
```
Comment (vmarchaud) [inline src/pmx.ts]: should check null

File: src/features/actions.ts [hunk 2]
```diff
@@ -50,29 +47,10 @@ export default class ActionsFeature implements Feature {
...
-      }
-
-      /**
-       * If data is an object == v2 protocol
-       * Pass the opts as first argument
-       */
-      if (typeof(data) === 'object' && data.msg === actionName) {
-        return fn(data.opts, reply)
+      if (fn.length === 2) {
+        return fn(data.opts || {}, reply)
```
Comment (vmarchaud) [inline src/features/actions.ts]: verify data is undefined

File: src/features/actions.ts [hunk 3]
```diff
@@ -133,8 +102,17 @@ export default class ActionsFeature implements Feature {
     }
   }
 
-  init (conf?, force?): Object {
+  initListener () {
+    if (this.listenerInitiated) return
+    if (ServiceManager.get('transport').transport) {
+      ServiceManager.get('transport').transport.on('trigger:*', this.listener)
```
Comment (vmarchaud) [inline src/features/actions.ts]: @wallet77 Do we need to bind `this.listener` to `this` here as @Unitech told before ?

File: src/features/actions.ts [hunk 4]
```diff
@@ -133,8 +102,17 @@ export default class ActionsFeature implements Feature {
...
+    } else {
+      process.on('message', this.listener)
```
Comment (vmarchaud) [inline src/features/actions.ts]: same here

File: src/services/transport.ts [hunk 5]
```diff
@@ -0,0 +1,112 @@
+import Debug from 'debug'
+import * as stringify from 'json-stringify-safe'
+
+const debug = Debug('axm:transportService')
+
+class TransportConfig {
```
Comment (vmarchaud) [inline src/services/transport.ts]: export every class definition

File: src/services/transport.ts [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+  private process: Process
+  private isStandalone: Boolean
+
+  init () {
+    this.isStandalone = false
+  }
+
+  initStandalone (config: TransportConfig, cb: Function) {
+    this.isStandalone = true
+    const Agent = require('/Users/valentin/Work/Keymetrics/pm2-io-agent-node')
```
Comment (vmarchaud) [inline src/services/transport.ts]: path

File: src/services/transport.ts [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+      axm_monitor: {}
+    }
+    this.agent = new Agent({
+      publicKey: this.config.publicKey,
+      secretKey: this.config.secretKey,
+      appName: this.config.appName
+    }, this.process, (err) => {
+      if (err) {
+        console.error(err)
+        process.exit(1)
```
Comment (vmarchaud) [inline src/services/transport.ts]: add return

File: src/services/transport.ts [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+      }
+      this.transport = this.agent.transport
+      debug('Agent launched')
+      return cb()
+    })
+  }
+
+  setMetrics (metrics) {
+    if (this.isStandalone) return this.process.axm_monitor = metrics
```
Comment (vmarchaud) [inline src/services/transport.ts]: use if/else

File: src/services/transport.ts [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+    this.send('axm:monitor', metrics)
+  }
+
+  addAction (action) {
+    debug(`Add action: ${action.action_name}:${action.action_type}`)
+    if (this.isStandalone) return this.process.axm_actions.push(action)
```
Comment (vmarchaud) [inline src/services/transport.ts]: use if/else

File: src/services/transport.ts [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+    return this.send('axm:action', action)
+  }
+
+  setOptions (options) {
+    debug(`Set options: [${Object.keys(options).join(',')}]`)
+    if (this.isStandalone) return this.process.axm_options = Object.assign(this.process.axm_options, options)
```
Comment (vmarchaud) [inline src/services/transport.ts]: use if/else

File: src/services/transport.ts [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+      process.send(JSON.parse(stringify({
+        type: channel,
+        data: payload
+      })))
+    } catch (e) {
+      debug('Process disconnected from parent !')
+      debug(e.stack || e)
+      process.exit(1)
+    }
+    return 0
```
Comment (vmarchaud) [inline src/services/transport.ts]: i believe this `return 0` is useless

**Comment by Eywek** [comment]
TODO: publish pm2-io-agent-node and require it

**Comment by Eywek** [comment]
TODO: Test exceptions/profiling with backend

File: src/features/actions.ts (hunk 3 above)
Comment (wallet77) [reply src/features/actions.ts]: If I remember it was an attempt to fix another issue but it didn't solve it.
I think it was committed but not necessary.

Comment (vmarchaud) [reply src/features/actions.ts]: Okay thanks, then we can go forward with this PR

**Comment by Eywek** [comment]
Can i merge this into development @wallet77 ?

**Comment by wallet77** [comment]
> Can i merge this into development @wallet77 ?

Yes but you should merge into master, development was created by Unitech for a big refactoring.

**Pull request merged**